src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java

```diff
@@ -0,0 +1,218 @@
+ * <a href="https://docs.oracle.com/en/java/javase/14/docs/specs/javadoc/doc-comment-spec.html#leading-asterisks">
+ * leading asterisks</a> in a Javadoc comment. The Check ensures that leading asterisks
+ * are aligned vertically under the second asterisk ( &#42; )
+ * of opening Javadoc tag. The alignment of closing Javadoc tag ( &#42;/ ) is also checked.
+ * If a line contains a non-whitespace character before the leading asterisk then that line will be
+ * ignored. Same applies for closing Javadoc tag.
+ * If the ending javadoc line contains a leading asterisk, then that leading asterisk's alignment
+ * will be considered, the closing Javadoc tag will be ignored.
+ * </p>
+ * <p>
+ * If you're using tabs then specify the the tab width in the
+ * <a href="https://checkstyle.org/config.html#tabWidth">tabWidth</a> property.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code violateExecutionOnNonTightHtml} - Control when to print violations if the
+ * Javadoc being examined by this check violates the tight html rules defined at
+ * <a href="https://checkstyle.org/writingjavadocchecks.html#Tight-HTML_rules">Tight-HTML Rules</a>.
+ * Type is {@code boolean}.
+ * Default value is {@code false}.
+ * </li>
+ * </ul>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code javadoc.asterisk.indentation}
+ * </li>
+ * <li>
+ * {@code javadoc.missed.html.close}
+ * </li>
+ * <li>
+ * {@code javadoc.parse.rule.error}
+ * </li>
+ * <li>
+ * {@code javadoc.unclosedHtml}
+ * </li>
+ * <li>
+ * {@code javadoc.wrong.singleton.html.tag}
+ * </li>
+ * </ul>
+ *
+ * @since 10.17.0
+ */
+@GlobalStatefulCheck
+public class JavadocLeadingAsteriskAlignCheck extends AbstractJavadocCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "javadoc.asterisk.indentation";
+
+    /** Specifies the column number of starting block of the javadoc comment with tabs expanded. */
+    private int commentBeginColNumber;
+
+    /** Specifies the line number of starting block of the javadoc comment. */
+    private int commentBeginLineNumber;
+
+    /**
+     * Specifies the column number of the leading asterisk
+     * where tabs are converted to a single whitespace.
+     */
+    private int singleSpaceAsteriskColNumber;
+
+    /** Specifies the lines of the file being processed. */
+    private String[] fileLines;
+
+    @Override
+    public int[] getDefaultJavadocTokens() {
+        return new int[] {
+            JavadocTokenTypes.LEADING_ASTERISK,
+        };
+    }
+
+    @Override
+    public int[] getRequiredJavadocTokens() {
+        return getAcceptableJavadocTokens();
+    }
+
+    @Override
+    public void beginJavadocTree(DetailNode rootAst) {
+        // this method processes and sets information of starting javadoc tag.
+        fileLines = getLines();
```

Comment:
As I mentioned at https://github.com/checkstyle/checkstyle/pull/14643#issuecomment-2077605335, and @Zopsss replied at https://github.com/checkstyle/checkstyle/pull/14643#issuecomment-2077760060, we get the whitespace preceding a given `*` for free with the LEADING_ASTERISK token. Why can't we use that instead of getLines?

Reply:
No we cannot use that. Here is why:

`getLines()` is used in two places, first to get the line of starting block ( /** ). We cannot use simple `rootAst.getText()` as it returns `JAVADOC` as its return value. So we use the `getLines()` to get the starting block line. Then we calculate its column number.

The second place is to get the line of ending block ( */ ). This block is not considered as a part of the javadoc ast, so to get the ending block we're using following code:

https://github.com/checkstyle/checkstyle/blob/d670bc3bc382150dd022907fec3239f693e296b4/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L144

Now if we use `javdocEndToken.getText()` to get the whole text of the ending line, the `getText()` method here always returns only `*/`, it ignores all the preceding whitespaces & any other preceding text and returns only the `*/`. So we're using the `getLines()` here to get the full ending javadoc line.

Though this does not happen in case of `LEADING_ASTERISK` token, for this token the `getText()` includes all the preceding whitespaces. So for this token we dont need to use `getLines()`. The `getText()` is behaving differently for different tokens maybe this is because of the implementation of our API. 


I've made the `fileLines` a global variable which stores the result of `getLines()` so we don't need to invoke it twice.

You might find this confusing and weird but I did not find any other easy way to get the things done.

Reply:
@Zopsss , https://checkstyle.org/writingjavadocchecks.html#Tool_to_print_Javadoc_tree_structure nodes above javadoc nodes does have all columns.
Please share AST trees that you have problems with and point more explicitly to a problem.

Reply:
Sure, let me elaborate the problem using an example:

#### JavadocTest.java
```java
public class JavadocTest {
  /**
      * Javadoc for foo.
    Testing */
  public void foo() {}
}
```

#### Javadoc AST:
```
-> java -jar ../checkstyle-10.14.2-all.jar -J JavadocTest.java
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
`--CLASS_DEF -> CLASS_DEF [1:0]
    |--MODIFIERS -> MODIFIERS [1:0]
    |   `--LITERAL_PUBLIC -> public [1:0]
    |--LITERAL_CLASS -> class [1:7]
    |--IDENT -> JavadocTest [1:13]
    `--OBJBLOCK -> OBJBLOCK [1:25]
        |--LCURLY -> { [1:25]
        |--METHOD_DEF -> METHOD_DEF [5:2]
        |   |--MODIFIERS -> MODIFIERS [5:2]
        |   |   |--BLOCK_COMMENT_BEGIN -> /* [2:2]
        |   |   |   |--COMMENT_CONTENT -> *\r\n      * Javadoc for foo.\r\n    Testing  [2:4]
        |   |   |   |   `--JAVADOC -> JAVADOC [2:5]
        |   |   |   |       |--NEWLINE -> \r\n [2:5]
        |   |   |   |       |--LEADING_ASTERISK ->       * [3:0]
        |   |   |   |       |--TEXT ->  Javadoc for foo. [3:7]
        |   |   |   |       |--NEWLINE -> \r\n [3:24]
        |   |   |   |       |--TEXT ->     Testing  [4:0]
        |   |   |   |       `--EOF -> <EOF> [4:12]
        |   |   |   `--BLOCK_COMMENT_END -> */ [4:12]
        |   |   `--LITERAL_PUBLIC -> public [5:2]
        |   |--TYPE -> TYPE [5:9]
        |   |   `--LITERAL_VOID -> void [5:9]
        |   |--IDENT -> foo [5:14]
        |   |--LPAREN -> ( [5:17]
        |   |--PARAMETERS -> PARAMETERS [5:18]
        |   |--RPAREN -> ) [5:18]
        |   `--SLIST -> { [5:20]
        |       `--RCURLY -> } [5:21]
        `--RCURLY -> } [6:0]
```

AST is showing 0 as column number for leading asterisk at line 3:

```
      * Javadoc for foo.
```

```
|--LEADING_ASTERISK ->       * [3:0]
```

But the actual column number of the leading asterisk is 7.

Now the AST is showing column 12 for the ending block of javadoc:

```
    Testing */
```

```
`--BLOCK_COMMENT_END -> */ [4:12]
```

The actual column number is 13 maybe it is showing 12 because of 0 base indexing but that's not our concern here. Our concern is that we don't need to check the alignment if there is some text before the ending block. Here we are getting the column number 12 but we cannot use that, we want column number as -1 if there is preceding text before the ending block based on our implementation. 

So to get the column number of the leading asterisk & ending block, I created a custom method called [getAsteriskColumnNumber()](https://github.com/checkstyle/checkstyle/blob/d670bc3bc382150dd022907fec3239f693e296b4/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L180). This method returns the leading asterisk's column number and for ending block it returns -1 if there is preceding text before ending block or ending block's column number if there isn't any preceding text.

For the examples I gave you above, the `getAsteriskColumnNumber()` would return 7 for the following:

```
      * Javadoc for foo.
```

which is the correct column number of leading asterisk.

It will return -1 for the following:


 ```
    Testing */
```

as there is preceding text before ending block.

---

The `getAsteriskColumnNumber()` is used inside a method `getAsteriskColumnNumberWithTabsExpanded()` which first gets the column number of the leading asterisk or the ending block, then gets the expanded column number of them. These expanded column numbers are then used to check the alignment.

Reply:
@romani @nrmancuso